Mathlib/LinearAlgebra/FreeModule/Int.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2024 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import Mathlib.Data.ZMod.Quotient
+import Mathlib.GroupTheory.Index
+import Mathlib.LinearAlgebra.FreeModule.PID
+
+/-! # Index of submodules of free ℤ-modules (considered as an `AddSubgroup`).
+
+This file provides lemmas about when a submodule of a free ℤ-module is a subgroup of finite
+index.
+
+-/
+
+
+@[simp] lemma Ideal.span_singleton_toAddSubgroup_eq_zmultiples (a : ℤ) :
+   (Ideal.span {a}).toAddSubgroup = AddSubgroup.zmultiples a := by
+  ext i
+  simp [Ideal.mem_span_singleton', AddSubgroup.mem_zmultiples_iff]
+
+namespace Basis.SmithNormalForm
+
+/-- Given a submodule `N` in Smith normal form of a free `R`-module, its index as an additive
+subgroup is infinite (represented as zero) if the submodule basis has fewer vectors than the basis
+for the module, and otherwise equals the product of the indexes of the ideals generated by each
+basis vector. -/
+lemma toAddSubgroup_index_eq {ι R M : Type*} {n : ℕ} [Fintype ι] [Infinite R] [CommRing R]
+    [AddCommGroup M] [Module R M] {N : Submodule R M} (snf : Basis.SmithNormalForm N ι n) :
+    N.toAddSubgroup.index = (if n = Fintype.card ι then
+      ∏ i : Fin n, (Ideal.span {snf.a i}).toAddSubgroup.index else 0) := by
+  classical
+  rcases snf with ⟨bM, bN, f, a, snf⟩
+  dsimp only
+  let N' : Submodule R (ι → R) := N.map bM.equivFun
+  let bN' : Basis (Fin n) R N' := bN.map (bM.equivFun.submoduleMap N)
+  have snf' : ∀ i, (bN' i : ι → R) = Pi.single (f i) (a i) := by
+    intro i
+    simp only [map_apply, bN']
+    erw [LinearEquiv.submoduleMap_apply]
+    simp only [equivFun_apply, snf, map_smul, repr_self, Finsupp.single_eq_pi_single]
+    ext j
+    simp [Pi.single_apply]
+  have hNN' : N.toAddSubgroup.index = N'.toAddSubgroup.index := by
+    let e : (ι → R) ≃+ M := bM.equivFun.symm
+    let e' : (ι → R) →+ M := e
+    have he' : Function.Surjective e' := e.surjective
+    simp only [N']
+    erw [Submodule.map_toAddSubgroup]
+    convert (AddSubgroup.index_comap_of_surjective N.toAddSubgroup he').symm using 2
+    simp only [e']
+    rw [AddSubgroup.comap_equiv_eq_map_symm]
+    rfl
```

Comment:
```suggestion
  have hNN' : N.toAddSubgroup.index = N'.toAddSubgroup.index := by
    set e : (ι → R) ≃+ M := ↑bM.equivFun.symm with he
    let e' : (ι → R) →+ M := e
    have he' : Function.Surjective e' := e.surjective
    convert (AddSubgroup.index_comap_of_surjective N.toAddSubgroup he').symm using 2
    rw [AddSubgroup.comap_equiv_eq_map_symm, he, hN', LinearEquiv.coe_toAddEquiv_symm,
    AddEquiv.symm_symm]
    exact Submodule.map_toAddSubgroup ..
```
The point of this proof is to avoid the final `rfl`, and replace it with a couple of `rw` (this makes the proof more robust). Note that I use the missing result
```lean
lemma LinearEquiv.coe_toAddEquiv_symm {R M N : Type*} [CommRing R] [AddCommGroup M] [Module R M]
    [AddCommGroup N] [Module R N] (f : M ≃ₗ[R] N) : (f.symm : N ≃+ M) = (f : M ≃+ N).symm := by
  rfl
```
that should be added somewhere in `Mathlib.Algebra.Module.Equiv.Defs`.

Reply:
I've filed #18965 for my `span_singleton_toAddSubgroup_eq_zmultiples`. Could you PR your `LinearEquiv.coe_toAddEquiv_symm` in a separate PR?

Reply:
I think you can safely add it to this PR

Reply:
OK, I've added it (in semilinear form).